src/lib/loaders/batchLoader.ts

```diff
@@ -0,0 +1,111 @@
+import DataLoader from "dataloader"
+import { chain, flatten, chunk } from "lodash"
+import config from "config"
+
+const { ENABLE_RESOLVER_BATCHING } = config
+
+/**
+ * This is just used to compare options sent into the dataloader.
+ * If the options are an object this stingifies the object in a way
+ * that they can be grouped as similar.
+ */
+const renderParams = key => {
+  if (typeof key === "string") {
+    return ""
+  }
+  const { id, ...params } = key
+  return Object.entries(params)
+    .map(entry => entry.join("="))
+    .sort()
+    .join("&")
+}
+
+interface GroupKeysResult {
+  id: string[]
+  size: number
+  [key: string]: any
+}
+export const groupKeys = (requestedKeys: string | { id }): GroupKeysResult[] =>
+  (chain(requestedKeys)
+    .groupBy(renderParams)
+    .values()
+    .map(values => chunk(values, 20))
+    .flatten()
+    .map((keys: string[] | { id }[]) => {
+      if (typeof keys[0] === "string") {
+        return { id: keys, size: keys.length }
+      }
+      return {
+        // @ts-ignore
+        ...keys[0],
+        id: keys.map(k => k.id),
+        size: keys.length,
+      }
+    })
+    .value() as unknown) as GroupKeysResult[]
+
+interface BatchLoaderArgs {
+  /**
+   * For resolvers designed to hit a list endpoint
+   * leave this argument empty. List endpoints
+   * support filtering options that single resource
+   * endpoints do not.
+   */
+  singleLoader?: any
+  multipleLoader: any
+  defaultResult?: any
+}
+
+export const batchLoader = ({
+  singleLoader,
+  multipleLoader,
+  defaultResult = null,
+}: BatchLoaderArgs) => {
+  if (!ENABLE_RESOLVER_BATCHING) {
+    return singleLoader ? singleLoader : multipleLoader
+  }
+  const dl = new DataLoader(keys => {
+    let groupedKeys = groupKeys(keys as any)
+
+    return Promise.all(
+      groupedKeys.map(keys => {
+        if (keys.id.length === 1 && singleLoader) {
+          return singleLoader(keys.id[0])
+        } else {
+          return multipleLoader(keys)
+        }
+      })
+    ).then(data => {
+      const normalizedResults = data.map((queriedGroup, groupIndex) => {
+        return groupedKeys[groupIndex].id.map(
+          id => queriedGroup.find(r => r._id === id) || defaultResult
```

Comment:
@joeyAghion I don't think it's obvious what's happening here, but I probably should explain it so that you can correct my assumptions if they're invalid. 

Let's say a query makes a request for three sale ids `sale-1, sale-2, sale-3` which results in a call like `/sale?id[]=sale-1&id[]=sale-2&id[]=sale-3`. The results will be an array of sales `[{ _id: sale-1, ... }, ...]`. It's important to note here that the sale id that we pass to the initial dataloader maps over to `_id` in the response object from gravity. At this point in the code, we're actually grouping and reordering results. So I'm taking the initial input ids (that the dataloader batched together) and ordering the results from gravity to match that sequence. 

So `['sale-1', 'sale-2', ...]` -> `{ _id: 'sale-1', ...}, { _id: 'sale-2', ...}, ...`. This is important because sometimes some (or all) of the calls to gravity might result in no data... but the dataloader expects to get a list of results the same length as the input (and in the same order). So the `|| defaultResult` part uses whatever we consider the fallback value for an entry in this list. In most cases that's just something like `null`. 

There could be a corner case here where an input is considered non-nullable or has some other default value... that should likely just be handled in the resolver though. 

Let me know if that makes any sense at all and if you have more questions.

Reply:
Might be worth adding a comment?

Reply:
Most Gravity APIs work when passed a true database identifier _or_ a "slug" (e.g., `/sale/1234` or `/sale/christies-sale`) and I believe many clients use a mix of these. In the specific case you're working on, I'd guess that we're dealing with real ids because they were probably sourced from artworks' `sale_ids` field. These end up as the `_id` property in response data, while the "slugs" appear in the `id` property.

Another consideration, if/when we do want to batch slug-based requests, is that many models support a whole "history" of slugs. E.g., you can request artist `frank-smith` and get back an object with `_id: 1234, id: "franklin-smith"` (i.e., the requested id appears nowhere in the response). For that reason, maybe we should just punt on supporting slugs in batching for now. If we care to support it at a later point, we might need to update APIs to respect the order of ids provided in params.